ignite/metrics/metric.py

```diff
@@ -56,6 +77,31 @@ def compute(self):
         """
         pass
 
+    def _sync_all_reduce(self, tensor):
+        if not (torch.distributed.is_available() and torch.distributed.is_initialized()):
+            # Nothing to reduce
+            return tensor
+
+        tensor_to_number = False
+        if isinstance(tensor, numbers.Number):
+            tensor = torch.tensor(tensor, device=self._device)
+            tensor_to_number = True
+
+        if isinstance(tensor, torch.Tensor):
+            # check if the tensor is at specified device
+            if tensor.device != self._device:
+                tensor = tensor.to(self._device)
```

Comment:
As far as I understood, if we have multiple gpus and call an op as reduce, than internally it collect data data from all gpus. In case if tensor is not defined on one of the gpus, the operation would hang...

So, in update we do not force all internal variables to be on the associated device (rank), but in compute it is necessary.